### PR TITLE
Ensure the original grad_mode is restored during backward

### LINKED ISSUE #31715: `enable_grad` context doesn't work as expected in backward function of torch.autograd.Function
## 🐛 Bug

This hasn't been an issue before, but it seems right now, it could be a serious blocking issue of one of my research projects. The gist is that a call to `torch.autograd.grad` seems to disable the `enable_grad` context in the `backward` function of `torch.autograd.Function`.

## To Reproduce
Here's a mini-repro:
```Python
import torch
from torch import nn


class BackwardBreaks(torch.autograd.Function):

    @staticmethod
    def forward(ctx, *args, **kwargs):
        assert len(args) == 3
        model = args[0]
        inp1, inp2 = args[1], args[2]
        ctx.model = model
        ctx.save_for_backward(inp1, inp2)
        return model.f(inp1) + model.g(inp2)

    @staticmethod
    def backward(ctx, *grad_outputs):
        model, inps = ctx.model, ctx.saved_tensors
        inp1, inp2 = inps
        grad_outputs, = grad_outputs
        with torch.enable_grad():
            print(torch._C.is_grad_enabled())
            grad1, = torch.autograd.grad(
                inp1 * 3, inp1, grad_outputs=grad_outputs
            )
            print(torch._C.is_grad_enabled())
            grad2, = torch.autograd.grad(
                inp2 ** 2, inp2, grad_outputs=grad_outputs
            )
        return None, grad1, grad2


class Model(nn.Module):
    def __init__(self):
        super().__init__()

    def f(self, inp):
        return inp * 2.

    def g(self, inp):
        return inp ** 2.


def backward_breaks_wrapper(model, inp1, inp2):
    return BackwardBreaks.apply(model, inp1, inp2)


def main():
    model = Model()
    inp1 = torch.randn(3, 1, requires_grad=True)
    inp2 = torch.randn(3, 1, requires_grad=True)

    result = backward_breaks_wrapper(model, inp1, inp2)
    result.backward(torch.ones(3, 1))
    print(inp1.grad)
    print(inp2.grad)


if __name__ == '__main__':
    main()
```

Run this w/ torch==1.3.0. I didn't try the bleeding edge version.

## Expected behavior
Clearly, we don't expect the output of `print(torch._C.is_grad_enabled())` to change. But afaik, the actual outputs here are `True` and then `False`.

## Environment
```
Collecting environment information...
PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Debian GNU/Linux rodete
GCC version: (Debian 8.3.0-6) 8.3.0
CMake version: version 3.13.4

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce GTX 1080
GPU 1: Quadro P1000

Nvidia driver version: 430.50
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.17.0
[pip] torch==1.3.1
[pip] torchdiffeq==0.0.1
[pip] torchfile==0.1.0
[pip] torchvision==0.4.2
[conda] torch                     1.3.1                    pypi_0    pypi
[conda] torchdiffeq               0.0.1                    pypi_0    pypi
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchsnooper              0.7.1                    pypi_0    pypi
[conda] torchvision               0.4.2                    pypi_0    pypi
```

## Additionals
Clearly, we could bypass this issue by setting `create_graph=True` in the first call to `torch.autograd.grad`. However, the other implications of this flag here isn't that great, considering that we're not aiming to obtain high-order derivs. 

Alternatively, we could use a new `torch.enable_grad` scope each time we decide to use `torch.autograd.grad`. Though this approach is less clean, and it's still nice to figure out why a call to `grad` somehow disable the context. Note, this doesn't happen if it's not in the backward function. 

Moreover, there seems to other bugs (which I haven't stably been able to get a mini-repro) in this backward function that's causing some non-deterministic segfaults and runtime errors even after I fix both the seed of torch and numpy. 

cc @ezyang @SsnL @albanD @zou3519 @gqchen

### PR DESCRIPTION
**Pull request by albanD**
Fix #31715

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31884 Ensure the original grad_mode is restored during backward**

Differential Revision: [D19301076](https://our.internmc.facebook.com/intern/diff/D19301076)

### CODE REVIEW

File: test/test_autograd.py [hunk 1]
```diff
@@ -3351,6 +3351,30 @@ def test_checkpointing(self):
...
+                        foo ** 3, foo, grad_outputs=go
+                    )
+                    self.assertTrue(torch._C.is_grad_enabled())
+                self.assertTrue(torch._C.is_grad_enabled() == original)
+                return grad
+
+        inp = torch.rand(3, requires_grad=True)
+
+        MyFunction.apply(inp).sum().backward()
+        MyFunction.apply(inp).sum().backward(create_graph=True)
```
Comment (gchanan) [inline test/test_autograd.py]: I'm guessing this is true, but just to check: this covers both original is True and False?

**Review by gchanan — APPROVED** [review]

File: test/test_autograd.py (hunk 1 above)
Comment (albanD) [reply test/test_autograd.py]: Exactly.

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 35df89be:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=35df89bed6a1b3f1f8db0e92b03fe6cc1d2b2e67).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4158399) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=81297455) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=81297455))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
 
  git checkout -b <new-branch-name> 
 
HEAD is now at 35df89bed6 Update on "Ensure the original grad_mode is restored during backward" 
+ git reset --hard 35df89bed6a1b3f1f8db0e92b03fe6cc1d2b2e67 
HEAD is now at 35df89bed6 Update on "Ensure the original grad_mode is restored during backward" 
+ git merge --allow-unrelated-histories --no-edit --no-ff ca72df06ae2f90ec0243094d8127f1b190fcea5f 
Auto-merging torch/csrc/autograd/engine.cpp 
Auto-merging test/test_autograd.py 
CONFLICT (content): Merge conflict in test/test_autograd.py 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>
